internal/pkg/service/stream/storage/statistics/statistics.go

```diff
@@ -52,12 +54,46 @@ type Aggregated struct {
 	Total Value
 }
 
+// Add returns a sum of the value and the given second value.
+// If only one of the values is a reset value then it returns a non-reset value using subtraction.
+// Note that the value can't be negative, in case of an underflow the fields will be 0.
 func (v Value) Add(v2 Value) Value {
-	v.SlicesCount += v2.SlicesCount
-	v.RecordsCount += v2.RecordsCount
-	v.UncompressedSize += v2.UncompressedSize
-	v.CompressedSize += v2.CompressedSize
-	v.StagingSize += v2.StagingSize
+	if v.Reset != v2.Reset {
+		if v.Reset {
+			return v2.Add(v)
+		}
+		if v2.SlicesCount > v.SlicesCount {
+			v.SlicesCount = 0
+		} else {
+			v.SlicesCount -= v2.SlicesCount
+		}
+		if v2.RecordsCount > v.RecordsCount {
+			v.RecordsCount = 0
+		} else {
+			v.RecordsCount -= v2.RecordsCount
+		}
+		if v2.UncompressedSize > v.UncompressedSize {
+			v.UncompressedSize = 0
+		} else {
+			v.UncompressedSize -= v2.UncompressedSize
+		}
+		if v2.CompressedSize > v.CompressedSize {
+			v.CompressedSize = 0
+		} else {
+			v.CompressedSize -= v2.CompressedSize
+		}
+		if v2.StagingSize > v.StagingSize {
+			v.StagingSize = 0
+		} else {
+			v.StagingSize -= v2.StagingSize
+		}
```

Comment:
I would move this into function, as the substraction sign in `Add` function looks weird. I would rather see `Sub(v2 Value)` function that does that so it is obvious that `Add` calls `Sub`.